File: 백준/Silver/9012. 괄호/괄호.py
# 9012. 괄호 - 오늘 나를 지겹게 한 ^^
stack = []
def gal(text):
    for ch in text:
        if ch == '(':
            stack.append(ch)
        elif ch == ')':
            if len(stack) == 0:
                return 'NO'
            else:
                if stack[-1] == '(':
                    stack.pop()
                    continue
                else:
                    return 'NO'

    if len(stack) == 0:
        return 'YES'
    else:
        return 'NO'

import sys
T = int(sys.stdin.readline())
for _ in range(T):
    bracket = sys.stdin.readline()
    stack = []

    result = gal(bracket)
    print(result)

File: 백준/Silver/9012. 괄호/괄호.java
import java.util.Scanner;
import java.util.Stack;

public class Main {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int T = scanner.nextInt();
        scanner.nextLine();

        for (int i = 0; i < T; i++) {
            String bracket = scanner.nextLine();
            String result = checkBrackets(bracket);
            System.out.println(result);
        }
    }

    public static String checkBrackets(String text) {
        Stack<Character> stack = new Stack<>();

        for (char ch : text.toCharArray()) {
            if (ch == '(') {
                stack.push(ch);
            } else if (ch == ')') {
                if (stack.isEmpty()) {
                    return "NO";
                } else {
                    if (stack.peek() == '(') {
                        stack.pop();
                    } else {
                        return "NO";
                    }
                }
            }
        }

        return stack.isEmpty() ? "YES" : "NO";
    }
}
